src/Microsoft.AspNet.Server.Kestrel/Infrastructure/KestrelThread.cs

```diff
@@ -98,67 +79,53 @@ public Task PostAsync(Action<object> callback, object state)
             return tcs.Task;
         }
 
-        private void PostCloseHandle(Action<IntPtr> callback, IntPtr handle)
+        private void ThreadStart(object parameter)
         {
-            lock (_workSync)
-            {
-                _closeHandleAdding.Enqueue(new CloseHandle { Callback = callback, Handle = handle });
-            }
-            _post.Send();
+            var tcs = (TaskCompletionSource<int>)parameter;
+            SetupLoop(tcs);
+            RunLoop();
         }
 
-        private void ThreadStart(object parameter)
+        private void SetupLoop(TaskCompletionSource<int> tcs)
         {
-            var tcs = (TaskCompletionSource<int>)parameter;
             try
             {
-                _loop.Init(_engine.Libuv);
-                _post.Init(_loop, OnPost);
+                _loop = new UvLoopHandle();
+                _post = new UvAsyncHandle(_loop, OnPost);
                 tcs.SetResult(0);
             }
             catch (Exception ex)
             {
                 tcs.SetException(ex);
             }
+        }
 
-            try
+        private void RunLoop()
+        {
+            using (_loop)
+            // _post is disposed in the uv_walk callback
             {
-                var ran1 = _loop.Run();
-                if (_stopImmediate)
+                try
                 {
-                    // thread-abort form of exit, resources will be leaked
-                    return;
+                    _post.Reference();
+                    _loop.Run();
+
+                    _loop.Validate();
+                    UnsafeNativeMethods.uv_walk(
```

Comment:
This definitely isn't the way to kill signalr connections. I'm ok with this being a last resort. SignalR itself listens to the ApplicationStopping token and releases connections gracefully when that happens. 

I'm just wondering under what circumstances with the new code does this code need to run.